components/hab/src/command/pkg/uninstall.rs

```diff
@@ -0,0 +1,131 @@
+/// 2. We find all packages on the filesystem and convert them into a graph
+/// 3. We do a BFS on the graph to get the dependencies in order
+/// 4. We check if the specified package has any reverse deps
+///     4a. If there are, we throw an error
+///     4b. If not, we delete the package
+/// 5. For each dependency we check if there are any packages which depend on it
+///     5a. If there are, we skip it
+///     5b. If there are not, we delete it from disk and the graph
+///
+pub fn start(ui: &mut UI, ident: &PackageIdent, fs_root_path: &Path, dryrun: bool) -> Result<()> {
+    // 1.
+    let pkg_install = PackageInstall::load(ident, Some(fs_root_path))?;
+    let ident = pkg_install.ident();
+    ui.begin(format!("Uninstalling package {}", &ident))?;
+
+    // 2.
+    let package_path = hfs::pkg_root_path(Some(&fs_root_path));
+    let packages = all_packages(&package_path)?;
+    let package_installs = load(&fs_root_path, &packages)?;
+
+    let mut graph = PackageGraph::new();
+    graph.build(&package_installs)?;
+
+    // 3.
+    let deps = graph.ordered_deps(&ident);
+
+    // 4.
+    match graph.count_rdeps(&ident) {
+        None => (), // package not in graph - m
+        Some(0) => {
+            let install = find_install(&package_installs, &ident).unwrap();
+            maybe_delete(ui, &install, dryrun)?;
+            graph.remove(&ident);
+        }
+        Some(c) => {
+            return Err(Error::CannotRemovePackage(ident.clone(), c));
+        }
+    }
+
+    // 5.
+    let mut count = 0;
+    for p in &deps {
+        match graph.count_rdeps(&p) {
+            None => (), // package not in graph - maybe panic!
+            Some(0) => {
+                let install = find_install(&package_installs, &p).unwrap();
+                if let Ok(true) = maybe_delete(ui, &install, dryrun) {
+                    count = count + 1;
+                }
+                graph.remove(&p);
+            }
+            Some(c) => {
+                ui.warn(format!(
+                    "Skipping package {}. It is a dependency of {} packages",
+                    p.clone(),
+                    c
+                ))?;
+            }
+        }
+    }
+
+    if dryrun {
+
+    } else {
+        ui.end(format!(
+            "Deleted package {} and {} dependencies",
+            &ident, count
+        ))?;
+    }
+    Ok(())
+}
+
+// JC: TODO: O(n)
+fn find_install(packages: &Vec<PackageInstall>, ident: &PackageIdent) -> Option<PackageInstall> {
+    for p in packages {
+        if p.ident() == ident {
+            return Some(p.clone());
+        }
+    }
+    return None;
+}
+fn maybe_delete(ui: &mut UI, install: &PackageInstall, dryrun: bool) -> Result<bool> {
+    let ident = install.ident();
+    if dryrun {
+        ui.status(Status::DryRunDeleting, &ident)?;
+        return Ok(false);
+    }
+
+    ui.status(Status::Deleting, &ident)?;
+    let pkg_dir = install.installed_path();
+    fs::remove_dir_all(&pkg_dir)?;
```

Comment:
On linux at least, this isn't atomic so concurrent `hab` users might end up trying to execute a half-delete package, causing a confusing error.  While we obviously can't avoid such errors completely, we could rename and then delete for a little bit more safety here.